OpenShift Tests Extension Framework Initial (#1676)
openshift/enhancements: enhancements/testing/openshift-tests-extension.md

```

The "list" verb will not be provided a kubeconfig and must output tests based solely
on environment information. If no environment arguments are provided, all tests must be
listed.

Listings are formatted as JSONL with one test description per line. A full listing contains
zero or more test description objects. The following example shows an abbreviated JSONL listing
for three tests.

```python
{"name": "T1", }
{"name": "T2", }
{"name": "T3", }
```

A test description object contains a number of attributes to inform `openshift-tests`
discovery and planning. Note that the following annotated example would be encoded into a single
line of output in actual listing output.

```python
{

    # Base, human-readable test name. 
    "name": "openssl version compliance",

    # If a test name is updated at any time in the future,
    # originalName must report the original name of the 
    # testing logic. This allows component readiness
    # to display the human-readable version of the test
    # name while considering test runs across name changes.
    "originalName": "security version compliance",

    # Optional subcomponent information.

>     "subcomponent": "ui",

Author: Agree this doesn't have much value since you made `--component` first class in OTE.

<!-- TOC -->

# OpenShift Tests Extensions

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

Today all conformance tests for OpenShift are in the monolithic [openshift-tests]
(https://www.github.com/openshift/origin) binary. These tests run on all the possible
flavors of OpenShift (more than 50) and contribute signal to tools like the
Release Controller and Component Readiness. However, the high barrier to entry in
openshift-tests means that teams are either not writing tests, or are using their
own homegrown solutions that ran in only very narrow configurations.

This enhancement proposes a framework to allow allows external repositories to
contribute tests to openshift-tests' suites with extension binaries, keeping tests
colocated with the features they are testing. It defines a standardized interface
for test discovery, execution, and result aggregation, allowing decentralized

> contributions while maintaining centralized orchestration.

Reviewer: > allowing decentralized contributions while maintaining centralized orchestration.

Is there any impact in how the new tests are added to the conformance suite, such as `openshift/conformance`?

Author: Common suites can be extended by adding tests to component repositories and citing those suites as parents; e.g. https://github.com/openshift-eng/openshift-tests-extension/blob/6cbf1d70a4500e2483275aff3dc6a74afe9c6365/cmd/example-tests/main.go#L27-L31 . Once the binary is registered, openshift-tests will run the inherited tests as part of the associated conformance suite.

### Non-Goals

1. Fully decompose `openshift-tests` or decentralize test orchestration & reporting.

## Proposal

### Concepts

#### Component

A component is a part of a software product. Component information includes the product name,
component name, and the type of the component. A single invocation of an external binary
can only associate with a single component (e.g. it can only list tests for / execute tests
for a single component).

#### Test ID

A unique combination of Component (Product, Type, Component Name) and original test name. 
A Test ID should not be repeated in a test listing across all participating extensions. 
The Test ID representing a unit of testing logic should not change over time (even if the 
human-readable test name changes, the Test ID should remain consistent by using the 
original test name).

#### Test Environment

A Test ID references a unit of testing logic. That logic may pass in some environments,
fail in others (e.g. a test that passes in gcp but fails in aws), and be
intentionally skipped in still others. A Test Environment
describes relevant facets of the environment in which a test would run or was run, including

> configuration information.

Reviewer 2: How will this Test ID generated?

within their image) to a simple registry datastructure in github.com/openshift/origin.

##### Non-Payload Extension Binaries

For non-payload components, contributors must advertise their extension binary for `openshift-tests`
to discover. This is accomplished by creating an ImageStream / ImageStreamTag with a special label:

`testextension.redhat.io/component=<product-name>-<component-name>`

An annotation on the ImageStreamTag will then identify the binary to extract and (optional)
arguments to pass ahead of the extension interface verbs.
`testextension.redhat.io/binary=<binary-path>.gz [--argument=value]`

Well-behaved operators should gate the creation of this imagestream/label/annotation on the existence of the
`TestExtensionAdmission` custom resource definition (the code should not attempt to
inspect `TestExtensionAdmissions` instances -- just check for the CRD itself so that operators
don't liter clusters with ImageStreams only applicable for testing in a production environment).

Optional Operator authors must ensure that the image carrying the extension binary
is identified in their ClusterServiceVersion (CSV) so that tools like `oc-mirror`

> will copy image(s) bearing extension binaries to disconnected clusters.

Reviewer 3: We also must ensure that all images used by the extension e2e tests are mirrored by `./openshift-tests images --upstream --to-repository=xyz`.

I could be wrong here, I think this list of images seems to be currently hardwired in openshift-tests binary during build. There is already some wording about adding new images to the list [here](https://github.com/openshift/origin/tree/master/test/extended/util/image), it will need to be either much stricter or we would need to get the list of images from extension binaries too.

Reviewer 4: Justin and I discussed this at one point, but I guess the outcome didn't end up in the enhancement. I think the outcome of that was new images would just need a separate PR to origin to add it.  It's low frequency enough that it shouldn't be too disruptive.

We need to solve the helper problem, though.

> will copy image(s) bearing extension binaries to disconnected clusters.

Reviewer 3: (starting a separate thread). To get the name of a mirrored image, a test is suggested to call `github.com/openshift/origin/test/extended/util/image.LocationFor("my.source/image/location:versioned_tag")` [here](https://github.com/openshift/origin/blob/b573a5bc983db72afd094614f5f61f725b975948/test/extended/util/image/README.md?plain=1#L20). Does the extension need to import `openshift/origin`? That could be problematic, especially if the extension imports incompatible version of `k8s.io/kubernetes`.

Reviewer 4: This is something we'll need to solve for sure.  We want to minimize vendoring requirements in the component repos for sure.  I do not want anyone to have to vendor origin code.

Perhaps we could pass the image locations to each tests binary and have OTE provide a LocationFor helper.  It could be passed either as a CLI flag or a path to a JSON/YAML file.

for different environments. A test "environment" describes the attributes of
cluster under test (e.g. the cloud provider, the topology, etc).

The implementation should eliminate tests that are not appropriate for the
specified environment.

The arguments accepted by the verb may vary over time and will be
defined by the extension interface version implemented by the binary.

**Version 1**

```
$ extension-binary list tests 
Environment Information:
  --component     "default" or "<product>:<type>:<component>"
  --platform      The hardware or cloud platform ("aws", "gcp", "metal", ...).
  --network       The network of the target cluster ("ovn", "sdn"). 
  --upgrade       The upgrade that was performed prior to the test run ("micro", "minor").
  --topology      The target cluster topology ("ha", "microshift", ...).
  --architecture  The CPU architecture of the target cluster ("amd64", "arm64").
  --installer     The installer used to create the cluster ("ipi", "upi", "assisted", ...).
  --config        Multiple. Non-default component configuration in the environment.
  --fact          Multiple. Facts advertised by cluster components.

>   --version       "major.minor" version of target cluster. 

Reviewer 5: Can we select tests by labels? OTA qe have some cases which should be executed in a special Prow job. 
For example, I have a case like below:
```
	It("Specified case OTA-12345", Label("OTA-12345"), func() {
		Expect(false).To(BeTrue(), "test failed")
	})
```
Then I want to only run this case.
I tried to run the case like below, it works, but I want to make it more simple:
```
./example-tests run-test --names "[sig-testing] openshift-tests-extension Specified case OTA-12345"
```
Thanks.

// or dictionary maintained in their repository.
allSpecsSet.Walk(func(*ExtensionTestSpec) {
    // do stuff
})

// Finally, some or all of the specs are added for the component. 
componentExtension.AddSpecs(allSpecsSet...)

customSuite := extension.Suite{
    Name: "my-custom-suite",
    Parents: []string{
        "openshift/conformance/parallel",
    },
    Qualifiers: []string{
        // CEL expression which selects appropriate tests
        "(test.tags.mytag==\"myvalue\"", 
    }
}

// AddGlobalSuite adds a suite whose qualifiers will extend to all test binaries
componentExtension.AddGlobalSuite(customSuite)

// AddSuite adds a suite whose qualifiers are automatically modified to filter on
// source = this component.
componentExtension.AddSuite(customSuite)

// The first component in the NewExtensionRegistry argument list
// will be selected by --component default, but also
// selectable with --component openshift:payload:hyperkube . 
// Most authors will be writing tests for one component, so origin
// omitting an option for --component, unless instructed otherwise, will "just 
// work".
componentRegistry := e.NewRegistry()

> componentRegistry.Register(componentRegistry)

Reviewer 6: @jupierce is it `componentRegistry.Register(componentExtension)`? thanks

Author: It is! Nice catch.

I'll correct this, but for implementers, better information on the state of the API can be found here:
https://docs.google.com/document/d/1cFZj9QdzW8hbHc3H0Nce-2xrJMtpDJrwAse9H7hLiWk/edit?tab=t.0#heading=h.66y4kqbj468a is the source for details on the API. https://github.com/openshift-eng/openshift-tests-extension/blob/main/cmd/example-tests/main.go provides an example.

Reviewer 6: Thank you for the information!

        # two runs.
        "testHash": "deadbeef",

        # In the future, we may run tests several times against
        # the same cluster in order to aggregate
        # test results more quickly. We may want to 
        # analyze results including this dimension if, for example,
        # we want to check whether the first run is as consistent
        # as the following N.
        "run": 0,

        "job": {
            # Name of the job which drove the test
            "name": "...",
            # Where the job information can be reviewed.
            "url": "..."
        },

        # Attributes of the host running the tests (not the 
        # system under test) or cloud account used. This can 
        # help us statistically analyze whether one system 
        # involved in running the testing is introducing 
        # failures that another is not (e.g. because it has 
        # a flakey network).
        "testDriver": {
            "cluster": "build01",
            "arch": "arm64",
            # Cloud account in which the ephemeral cluster was 
            # installed.
            "account": "34908092383"
        }
    }
}
```

> ### Risks and Mitigations

Reviewer 3: With all tests in `openshift/origin`, we need to bump k8s.io/kubernetes with many useful e2e functions only there. When we move the tests to individual repos, all these repos will endure pain with updating k8s.io/kubernetes. Upstream does not even pretend it has a stable API and it often breaks.

Alternatively, we would need to spend some non-trivial time rewriting the tests not to depend on k8s.io/kubernetes/test/e2e/framework.

Reviewer 4: Would it be helpful to have a discussion about your particular use case? I might be missing some context. What e2e functions are you using from k8s.io/kubernetes/test/e2e/framework? Which tests are you looking to move to external binaries for which repos?

- Our next OTE customer after migrating k8s-tests-ext is ovn-kubernetes, and the ginkgo tests already exist in their repos to run them.  We'd mostly be running them unmodified from upstream

-  The other use case we looked at was moving QE's openshift-tests-private to the component repos, and those don't use anything from k8s.io/kubernetes/test/e2e/framework.

Reviewer 3: I understand the proposal that tests from origin should eventually move to  individual team repos. If I look what imports k/k in today's origin: ```
$ grep -R -l k8s.io/kubernetes/test/e2e/framework  test/extended  | wc -l
194
```

If we move these 194 files to individual team repos, these teams will either need to use some other test framwork than `k8s.io/kubernetes/test/e2e` or rebase unstable k/k regularly.

If the tests can stay in origin, then I'm fine.

Reviewer 4: There's no plan to plan to force anyone to move, but I'm also not entirely sure I understand the problem.  The vendoring of k8s.io/kubernetes/test/e2e/framework would just be in the component repo - there's no interdependency between origin's deps and an OTE binary's deps.

5. Maintain and increase testing coverage.
6. Maintain and increase code quality.
7. Maintain and improve the signal TRT provides through Component Readiness.
8. Provide a standard model for testing component configurations.
9. Allow distributed contribution of tests while introducing centralized mechanisms to require important tests from component owners and improve coverage over time.
10. Improve testing efficiency by reducing initialization time.
11. Allow layered products to participate in OpenShift testing & component readiness.

### Non-Goals
1. Fully decompose `openshift-tests` or decentralize test orchestration & reporting.

## Proposal

### Concepts

#### Component
A component is a part of a software product. Component information includes the product name, 
component name, and the type of the component. A single invocation of an external binary 
can only act on a single component.

#### Subcomponent
A further refinement of a component, if necessary. When combined with Component information,
it should be sufficient to file a Jira ticket that will reach domain experts. 

#### Test ID
A unique combination of Component, Subcomponent, and test name. A Test ID should not be

> repeated in a test listing across all participating extensions.

Author: I say that `testHash` is a digest of test IDs -- it is a method of checking for poorly interacting tests. Test ID is just a string that unique identifies a unit of test logic across time. I was thinking this would be `<product>-<component type>-<component>-<original test name>` . 
openshift/origin gets to come up with the actual format of the string they want to use based on the information returned by the tests extension framework implementation. 
The onus of maintaining the original test name as metadata in the extension is placed on the component owner.

with a `system:admin` kubeconfig when tests are run. Administrators
must opt-in to this risk by instantiating one or more `TestExtensionAdmission`
objects. 

`TestExtensionAdmission` contain a list of namespaces/imagestreams 
from which `openshift-tests` is permitted to extract extension binaries. This
can include a wildcard patterns for namespace, imagestream, or both.

```yaml
kind: TestExtensionAdmission
metadata:
  name: example
spec:
  permit:
  # All imagestreams in the openshift namespace with the testextension label
  - "openshift/*" 
  # All imagestreams on the cluster with the testextension label
  - "*/*" 
```

`openshift-tests` will SKIP a single synthetic test whenever an extension binary is detected in 
any imagestreamtag but which is not permitted for execution by a `TestExtensionAdmission` 
instance. This will allow users to detect tests that are available but not yet explicitly
permitted to run.

#### Binary Format
Extension binaries are extracted from container images into a running Pod executing 
`openshift-tests`. Contributors do not necessarily know which version of RHEL the
`openshift-tests` binary will be running on, so, for maximum portability, test

> extension binaries should be statically compiled.

Reviewer 4: Do we need to statically link openshift-tests too? It's not today AFAICT.

Reviewer 7: It's not and that's probably because FIPS tests generally need to run with dynamic linking, like most of what they're exercising. For other binaries that need to be "agnostic" they are generally compiled for both RHEL 8 and 9.

Not certain that these binaries would need to operate with FIPS enabled.

Author: The static linking only serves to insulate us from the issues which arise when binaries are copied from one environment to another. There is no need for openshift-tests to be statically linked for the purposes of CI since it will be running in its native container. Maybe it would be beneficial to end users who aren't building their own or running inside the `tests` container, but if we haven't received any complaints, it is probably not an issue.

> Not certain that these binaries would need to operate with FIPS enabled.

It wouldn't hurt. Being statically linked, extension binaries can't be FIPS compliant today, but we can prevent the "FIPS or Die" feature from killing them when run in a FIPS-enabled environment.

Author: https://github.com/openshift-eng/openshift-tests-extension/pull/13

```
info      - Output test contribution extension version and metadata.
list      - Output tests supported by this extension.
run-test  - Run one or more tests and output results.
config    - Component configuration management.
```

##### Info - Extension Metadata
The extension interface will evolve over time. For this reason, the `info` verb
will output information about which version of the interface the binary supports. The origin
framework will support some level of skew in the interface versions and invoke the binaries
with verbs & arguments consistent with the interface version they support.

Info also provides information to origin about how to construct new or contribute new
test suites.

Annotated example `info` output is provided below.
```python
{
    # The extension interface version the binary implements.
    # The details of the interface and version will usually be
    # unnecessary for a contributor to understand as they will
    # vendor in the majority of the implementation from the
    # "Test Extension Support Module" explained later.
    "apiVersion": "1",

    "source": {
        # The commit from which this binary from compiled.
        "commit": "87fdbaa"

>     },

Author: Many `make` flows use the current commit. It seems less common to burn your source repo into a binary. Maybe make it optional?

    # A single extension can carry tests for multiple
    # components. However, each invocation of the binary
    # can only represent and act on information relative
    # to a single component.
    # If a binary carries tests for multiple components,
    # it must register itself multiple times, with
    # different pre-verb arguments which openshift-tests
    # will provide verbatim with every invocation; e.g.
    # if the binary registers with "--component hyperkube",
    # verb invocations will be as follows:
    #    extension-binary --component hyperkube info
    #    extension-binary --component hyperkube list ...
    "component": {
        "product": "openshift",
        "type": "payload",
        "name": "hyperkube"

>     },

Reviewer 4: What's the value of an additional grouping of tests beyond suite? 

What if tests are cross-component? I guess we still ask there's a primary owner, I'm just wondering why we need the additional slicing of tests

Author: > Does this get exposed in the test name at all?
No. The test name is the simple, human readable string. Test ID, which openshift/origin derives from test metadata, is something like  `<product>-<component-type>-<component-name>-<original test name>" and should be unique across extensions and consistent across time (if the test name changes).

> What's the value of an additional grouping of tests beyond suite?
The intent is for non-OpenShift products to participate in the testing.  Component Readiness may provide views for different products.

    #              across all parallel extension binary invocations.
    # "bucket"   - avoids being called in the same planning bucket.
    # If two tests have a conflict name in common, the isolation
    # mode will apply. "*" is a special conflict name will 
    # ensures complete isolation in the requested mode (e.g.
    # guaranteeing the test will be the only one to
    # a given invocation of run-test if mode=instance).
    "isolation": {
      "mode": "exec",
      "conflict":  [
        "gpu",
      ]
    },
      
    # Tests that require significant resources can identify
    # that fact so that the planning algorithm will not
    # overdraw on allocatable pod resources for a set of
    # tests run in parallel.
    "memory": "1Gi",
    
    # If a duration can be roughly predicted, inclusion of this
    # information may improve the test execution 
    # planning algorithm performance.  
    "duration": "2s",
      
    # Timeout information may also support efficient
    # planning.
    "timeout": "16s",
  },

  # Tests can be identified as "informing" or "production".
  # Informing tests will not negatively impact default views

>   # in Component Readiness.

Author: I was worried about using "blocking" here because I immediately think "blocking release controller jobs." It's very possible to define a test that is mature (production), that won't run in a blocking job. I also toyed with: 
- alpha: I just want to see what this test looks like in CR.
- beta: I want this in a non-default view and raising Jiras on failure.
- production: I want this to be on the dashboard that decides releases. I could see calling this blocking because you are making a statement that broader than maturity. You are also stating how it should be used within component readiness.

>   # in Component Readiness.

Reviewer 4: I think I still prefer the consistency of the terminology: Informing tests won't fail an individual CI job, blocking tests do.  Informing jobs don't fail a payload, blocking jobs do.

>   # in Component Readiness.

Reviewer 8: I would lean towards informing+blocking for consistency, those terms are well established here. I can see how someone might occasionally get confused that it means the test runs in a blocking job,  but I think the alternative (correct) assumption meaning it can fail a job would be much more common.

  # However, policies may be established in the future that 
  # cause tests that remain informing too long to be treated
  # as if they are production.
  "lifecycle": "informing",
  
}
```

##### Run-Test - Running Extension Tests
The `run-test` verb is used to cause the extension binary to actually run 
discovered tests. `run-test` accepts test environment arguments as discussed
in the "list" verb section (in order to reduce fact discovery time needed
for each test run) and one or more test names to invoke.
```
$ extension-binary run-test 
  --platform      The hardware or cloud platform ("aws", "gcp", "metal", ...).
  ...other context arguments...
  --config        The component configuration profile to assume is active on the cluster.
  --name | -n     Test name to invoke (-n can be specified multiple times).
```

> `run-test` should run the specified tests in parallel. To prevent 

Reviewer 4: This is interesting.  We'll have to rework the parallelism logic on origin to support this, but should reduce the overhead of having to call these binaries individually for each test.

`openshift-tests` will randomize tests into multiple different parallel
executions of the extension binary in order to amortize initialization
costs.

Standard output from the invocation will be serialized into JSONL with
each line associated with a single test run result.

```python
{ "name": "T1", "result": "success", }
{ "name": "T1", "result": "success", }
{ "name": "T1", "result": "success", }
```

A single test result will contain the following information, encoded on a single line:
```python
{
    "name": "test name",
    
    # Duration of test run in milliseconds
    "duration": "1012",
    
    # The outcome of the test; pass, fail, skip, timeout.
    "result": "pass",
    "output": "standard output of the test",
    "error": "standard error of the test",
    
    # Human-readable messages to further explain 
    # skips / timeouts / etc.
    "messages": [
        "example message"
    ]
}
```

##### Config - Component Configuration Testing

A component can advertise that it wants to be exercised in multiple different configuration.
`openshift-tests` will plan an efficient method of testing those configurations and
ensure tests appropriate to that configuration are run.

> 

Reviewer 4: What happens when you run the same tests against multiple configurations? How do we report those results?

I think we need to describe how that works. A simple way to handle this would be make sure the results reported in the JUnit append `[Config:XYZ]` to the name.

Author: `Test Result Aggregation` touches on this and suggests we move beyond junit. A single test result is a rich structure including information about success/failure AND the configuration in which it was run.

        # parallelization and bucketing. Analyzing seeds against
        # one another should highlight parallelization issues.
        # Using the same seed with the same tests should result
        # in identical planning and parallelization. 
        "seed": 15,
        
        # A sha256 of the tests IDs which were run. This, plus
        # seed should ensure identical execution pattern between
        # two runs.
        "testHash": "deadbeef",
        
        # In the future, we may run tests several times against
        # the same cluster in order to aggregate
        # test results more quickly. We may want to 
        # analyze results including this dimension if, for example,
        # we want to check whether the first run is as consistent
        # as the following N.
        "run": 0
    }
}
```

### Risks and Mitigations

#### Binary Incompatibility
Copying a binary from a different image into a running pod exposes us to binary incompatibility. 

##### CPU Architecture 
An arm64 binary will not run in an amd64 based pod. `openshift-tests` will need to run on 
a build farm node with a CPU architecture consistent with architecture of the cluster 

> payload under test.

Reviewer 4: Or, use the multiarch payload

Author: This is harder to accomplish when dealing with CI payloads and pre-submit testing.

---
title: extended-platform-tests
authors:
  - "@jupierce"
  - "@stbenjam"
reviewers:
  - "@deads2k"
creation-date: 2024-09-05
last-updated: 2024-09-05
status: implementable
superseded-by:

>   - "enhancements/testing/openshif-tests-extension.md"

Author: As always, glad I missed the `t` and not the `f` :) . Removing this line altogether. Just a copy & paste carry-over I should have dropped.

10. Improve testing efficiency by reducing initialization time.

### Non-Goals
1. Fully decompose `openshift-tests` or decentralize test orchestration & reporting.

## Proposal

### Concepts

#### Component
A component is a part of a software product. Component information includes the product name, 
component name, and the type of the component. A single invocation of an external binary 
can only associate with a single component (e.g. it can only list tests for / execute tests
for a single component).

#### Subcomponent
A further refinement of a component, if necessary. When combined with Component information,
it should be sufficient to file a Jira ticket that will reach domain experts. 

#### Test ID
A unique combination of Component, Subcomponent, and test name. A Test ID should not be
repeated in a test listing across all participating extensions.

#### Test Environment
A Test ID references a unit of testing logic. That logic may pass in some environments,
fail in others (e.g. a test that passes in gcp but fails in aws), and be 
intentionally skipped in still overs. A Test Environment

> describes relevant facets of environment in which a test would run or was run, including 

Author: Thanks! Next push.

3. Discover tests available from those extensions (running `list` verb and parsing standard out).
4. Plan an appropriate and efficient testing strategy based on the framework invocation.
5. Execute the tests using subsequent invocations of the discovered binaries (`run-test` invocations).
6. Collecting the result and output of the test invocations and aggregating it into overall test suite results.

#### Binary Discovery
The existence of test extension binaries can be registered one of two ways by test authors. 

##### OpenShift Payload Extension Binaries
For OpenShift payload components contributors can advertise the existence of an extension binary
by adding information (the imagestream tag for the OCP payload component and the path to the binary 
within their image) to a simple registry datastructure in github.com/openshift/origin. 

##### Non-Payload Extension Binaries
For non-payload components, contributors must advertise their extension binary for `openshift-tests`
to discover. This is accomplished by creating an ImageStream / ImageStreamTag with a special label:

> `testextension.redhat.io/component=<product-name>-<component-name>`

Reviewer 7: I think you are talking about image labels here? Which on tagging get transcribed to an ImageStreamTag annotation? Are we talking about an imagestream populated on the CI cluster, or one included in the payload manifests?

Couple potential issues there:
1. ART ImageStreams don't tag non-payload images
2. Not all payloads are built from imagestreams... multi for instance is not. I suppose these could get the info directly from the image labels?

Author: I'm describing an ImageStream instantiated on the cluster under test / ephemeral cluster. The flow would be:
1. A CI job stands up an cluster to test.
2. The job instantiates a `TestExtensionAdmission` on the ephemeral cluster.
3. The job installs an optional operator that wants to participate in testing.
4. The optional operator (or CI job) creates an ImageStream and labels it according the described labeling pattern to advertise a tests extension binary.
5. openshift-tests runs, analyzes the cluster under test for ImageStreams advertising extension binaries, verifies they are permitted by `TestExtensionAdmission` instance(s) and extracts them.

    },

    # A single extension can carry tests for multiple
    # components. However, each invocation of the binary
    # can only represent and act on information relative
    # to a single component.
    # If a binary carries tests for multiple components,
    # it must register itself multiple times, with
    # different pre-verb arguments which openshift-tests
    # will provide verbatim with every invocation; e.g.
    # if the binary registers with "--component hyperkube",
    # verb invocations will be as follows:
    #    extension-binary --component hyperkube info
    #    extension-binary --component hyperkube list ...
    "component": {
        "product": "openshift",
        "type": "payload",
        "name": "hyperkube"
    },

    "configurations": [
        {
            "profile": "common-mode-1",
            "description": "Enables a common, non-default mode of operation where ....",
            "application": {
                # If applying the configuration implies a disruption, inform 
                # openshift-tests, so that it can be accounted for in overall
                # disruption reporting.
                "disruption": "1m",

>                 # If, after application the component configuration, openshift-tests

Reviewer 7: ```suggestion
                # If, after applying the component configuration, openshift-tests
```

              # that fact so that the planning algorithm will not
              # overdraw on allocatable pod resources for a set of
              # tests run in parallel.
              "memory": "1Gi",
              
              # If a duration can be roughly predicted, inclusion of this
              # information may improve the test execution 
              # planning algorithm performance.  
              "duration": "2s",
                
              # Timeout information may also support efficient
              # planning.
              "timeout": "16s",
            },
            
            
        }  
    ],
    
    # Additional suites the extension wants to advertise / participate in.
    "suites": [
        {
            # Here, the extension is advertising a new suite, unknown
            # to the origin framework until discovered through the
            # binary. 
            "name": "fips/conformance",

            "parents": [

                # This suite can be run separately or will be included
                # automatically in other suites known to origin.
                # Here, the extension is informing origin that the suite

>                 # is a subuite of openshift/conformance. All information

Reviewer 7: sub-suite?
```suggestion
                # is a sub-suite of openshift/conformance. All information
```

                # about a suite is additive across extension binaries.

>                 # For example, mulitple binaries can advertise the same

Reviewer 7: ```suggestion
                # For example, multiple binaries can advertise the same
```

#### Runtime Size / Speed
Statically linked binaries are large relative to dynamically linked binaries. As the number
of extension binaries increase, so will the required tmpfs space used by pods running the
tests as those binaries are stored locally within the pod.

To mitigate this risk, we need to be aware of it, overprovision test pods with respect to 
the memory normal test execution requires, and also include a test failure that will 
TRT attention if the pod is using > 80% of allocatable memory.

#### Image Size
Statically linked binaries are relatively large. As the number of test extension binaries
increases, it may lead to a noticeable increase in overall payload size.

#### Poor Extension Implementation
Providing a contract that must be implemented without by test extensions without centralized 
control & review of the implementation leaves space for contributions that violate the 
expectations of that contract.

### Version Skew Strategy

The majority of test extension binaries will reside within the OpenShift payload under test, and thus cannot
skew. For non-payload components, test filtering, performed by the extension binary itself can take the current

> platform version into account when determining appropriate tests to run.

Reviewer 4: One problem that could arise is the openshift-tests binary being used could be newer or older than the payload being run on the cluster under test.  We might want to try to detect this and warn/error.

One implication: imagine a cluster running 4.17 and then using a 4.18 openshift-tests that expects to find an external test binary in the "ovn-kubernetes" image, and it doesn't.

Author: Good call - We should code for that implication. If openshift-tests and the cluster under test differ in minor version, failure to extract an external binary from the registry should probably turn from an error to a warning.

OpenShift test suites.
// Layered on commits from Stephen with good description from gdoc if we move forward.

## Motivation

Increase the number of tests in the OpenShift Tests Suite by reducing the
effort, complexity, and risk of introducing new tests.

### Goals

1. Reduce the effort, complexity, and risk of introducing new tests to OpenShift.
2. Allow tests to be introduced in the same pull request as the PR which fixes a bug or introduces a feature.
3. Allow component owners to introduce new test suites / add tests to existing suites.
4. Allow non-OpenShift components, including layered products, to participate in extending OpenShift tests and Component Readiness.
5. Maintain and increase testing coverage.
6. Maintain and increase code quality.
7. Maintain and improve the signal TRT provides through Component Readiness.
8. Provide a standard model for testing component configurations.
9. Allow distributed contribution of tests while introducing centralized mechanisms to require important tests from component owners and improve coverage over time.
10. Improve testing efficiency by reducing initialization time.

### Non-Goals

> 1. Fully decompose `openshift-tests` or decentralize test orchestration & reporting.

Reviewer 8: Might be worth noting here this is purely for e2e ginko tests, and "monitortests" which are instantiated at the start, watch the cluster, and generate junit results at the end based on what it observed, will remain in origin.

Author: @dgoodwin - I've added "list monitors" and "run-monitor" if you want to take a look.

        "source_url": "github.com/openshift/kubernetes"
    },

    # A single extension can carry tests for multiple
    # components. However, each invocation of the binary
    # can only represent and act on information relative
    # to a single component.
    # If a binary carries tests for multiple components,
    # it must register itself multiple times, with
    # different pre-verb arguments which openshift-tests
    # will provide verbatim with every invocation; e.g.
    # if the binary registers with "--component hyperkube",
    # verb invocations will be as follows:
    #    extension-binary --component hyperkube info
    #    extension-binary --component hyperkube list ...
    "component": {
        "product": "openshift",
        "type": "payload",
        "name": "hyperkube"
    },

    "configurations": [
        {
            "profile": "common-mode-1",
            "description": "Enables a common, non-default mode of operation where ....",
            "application": {
                # If applying the configuration implies a disruption, inform 
                # openshift-tests, so that it can be accounted for in overall
                # disruption reporting.

>                 "disruption": "1m",

Reviewer 8: This makes me a little nervous if it's happening in other repos and we didn't have good visibility into people abusing this because a problem popped up. Probably not a core concern for this enhancement though.

  --config        The component configuration to assume is active on the cluster.
  --version       "major.minor" version of target cluster. 
```

The "list" verb will not be provided a kubeconfig and must output tests based solely
on environment information. If no environment arguments are provided, all tests must be listed.

Listings are formatted as JSONL with one test description per line. A full listing contains
zero or more test description objects. The following example shows an abbreviated JSONL listing 
for three tests.
```python
{ "name": "T1", }
{ "name": "T2", }
{ "name": "T3", }
```

A test description object contains a number of attributes to inform `openshift-tests`
discovery and planning. Note that the following annotated example would be encoded into a single 
line of output in actual listing output.
```python
{
  
  # Base, human-readable test name. 
  "name": "openssl version compliance",

  # If a test name is updated at any time in the future,
  # originalName must report the original name of the 
  # testing logic. This allows component readiness
  # to display the human-readable version of the test
  # name while considering test runs across name changes.

>   "originalName": "security version compliance",

Reviewer 8: I really like baking this into the component repos, might get us more renames as very very few people go to the mapping repo.

It would be neat if something could comment on PRs where we see added+removed tests that "if this was a rename, please do ...". That would be relatively often a rename. Problem for another day though.

    
    # The outcome of the test; pass, fail, skip, timeout.
    "result": "pass",
    "output": "standard output of the test",
    "error": "standard error of the test",
    
    # Human-readable messages to further explain 
    # skips / timeouts / etc.
    "messages": [
        "example message"
    ]
}
```

##### Config - Component Configuration Testing

A component can advertise that it wants to be exercised in multiple different configuration.
`openshift-tests` will plan an efficient method of testing those configurations and
ensure tests appropriate to that configuration are run.

The `config` verb will not return until the extension has watched the component
fully apply the configuration (or it should timeout with an error).
```
$ extension-binary config
  --profile <profile name>      The configuration profile to apply or "default".
```

The component configuration is considered part of the Environment and thus passed
in to `list` in order to determine appropriate tests for the configuration. 

Component authors may choose to reduce the number of tests run for non-default
configuration profiles, focusing only on tests likeliest to fail based on the 

> configuration change, in order to reduce overall execution time.

Reviewer 8: Could we get an example on this one, I couldn't come up with a use case immediately.

Author: Let's say you want to test a more verbose logging level configuration option in your CRD. You may have hundreds of tests that fully exercise your component in the 'default' configuration, but it is overkill to run them all again simply to verify that debug log statements are being emitted by your pod when verbose logging is enabled.
So you expose an extension configuration for the verbose logging level, and output only one test when asked for a list in that configuration. All that test does is read the component's pod logs and makes sure that it sees debug output.

If you expose a configuration that disables http/1, you might just want to run a test that verifies an http/1 connection is rejected and an http/2 connection is accepted. If you want to test branding, you might just want to verify that the HTML you scrape contains the newly configured name. If you expose a threshold, you might just want to test that a single expected alert is firing after configuring it.

I can add these to the doc if you buy the premise.

```python
{
    # Human-readable test name
    "name": "openssl security compliance",
    
    # Original human-readable test name, if it has changed over time.
    "originalName": "fips security compliance",
    
    # The Test ID openshift/origin has defined for this test based
    # on component and test metadata. It is meant to be unique
    # across all other components and consistent across time
    # (even if the human-readable name for a unit of test logic
    # is changed).
    "id": "openshift-payload-api-server-fips security compliance",
    
    "result": "pass",

    # ...other test result information...
    
    # Information necessary to fully qualify the 
    # test.
    "component": {
        "product": "openshift",
        "type": "payload",
        "component": "hyperkube",
        "subcomponent": "logs",
    },
    
    "environment": {
        "platform": "aws",
        "architecture": "amd64",
        # ...others...
        
        # Configuration information is also included. 
        "configuration": {
            # The configuration id applied to the component
            # before the test was run.
            "component": "default",
        }

>     },

Reviewer 8: Environment is a little unusual in the results for every test, that seems like mostly a characteristic of a job and a lot of duplication?

Author: There are fixed aspects of the environment, like GCP vs AWS, but the enhancement suggests that environment include component configuration information -- a single job being able to apply and test multiple different configuration options. Imagine an operator being able to cycle through several of its typical configurations, running the same tests or tests specific to those configurations, during the execution of a single job. That configuration is relevant to the outcome of a test and Component Readiness must be able to differentiate the same test name running in one configuration vs another.

A next question would be why we would store those static environmental aspects in the aggregated results file alongside each test. My hope there is that the results files can begin to stand alone. You can just push the file content into a database and you know everything you need to know from the resulting DB. You don't need parse prowjob job names, for example, to derive additional context about how the test was run. Many tools can ingest a comprehensive file like this directly, so our options for analysis expand. Imagine wanting to move to a new database or use local tooling to analyze the data. With a comprehensive file format, we just ingest the file into our target analysis tool -- no custom logic like what we have in the cloud function required to pull bits and pieces from multiple artifacts.

                
              # Timeout information may also support efficient
              # planning.
              "timeout": "16s",
            },
            
            
        }  
    ],
    
    # Additional suites the extension wants to advertise / participate in.
    "suites": [
        {
            # Here, the extension is advertising a new suite, unknown
            # to the origin framework until discovered through the
            # binary. 
            "name": "fips/conformance",

            "parents": [

                # This suite can be run separately or will be included
                # automatically in other suites known to origin.
                # Here, the extension is informing origin that the suite
                # is a sub-suite of openshift/conformance. All information
                # about a suite is additive across extension binaries.
                # For example, multiple binaries can advertise the same
                # suite -- if they advertise different parents, then
                # origin will treat the suite as a subset of each 
                # identified parent.
                "openshift/conformance"

>             ],

Reviewer 4: How are parents and children joined together?

In this example, is the suite named "openshift/conformance/fips/conformance"?

Author: As imagined here, `fips/conformance` is a suite that can be executed by origin in isolation -- the specified name is fully qualified. 
However, because it is claiming `openshift/conformance` as a parent, if origin runs `openshift/conformance`, tests from this suite definition will be included.

Reviewer 4: openshift-tests-extension (OTE) as written allows contributing tests to multiple suites - if we can do that, does we need the idea of parents?

    # "instance" - avoids being called along with a conflicting
    #              test in a single run-test invocation.
    # "exec"     - avoids being called at the same time in a
    #              across all parallel extension binary invocations.
    # "bucket"   - avoids being called in the same planning bucket.
    # If two tests have a conflict name in common, the isolation
    # mode will apply. "*" is a special conflict name will 
    # ensures complete isolation in the requested mode (e.g.
    # guaranteeing the test will be the only one to
    # a given invocation of run-test if mode=instance).
    "isolation": {
      "mode": "exec",
      "conflict":  [
        "gpu",
      ]
    },
      
    # Tests that require significant resources can identify
    # that fact so that the planning algorithm will not
    # overdraw on allocatable pod resources for a set of
    # tests run in parallel.
    "memory": "1Gi",
    
    # If a duration can be roughly predicted, inclusion of this
    # information may improve the test execution 
    # planning algorithm performance.  
    "duration": "2s",
      
    # Timeout information may also support efficient
    # planning.

>     "timeout": "16s",

Reviewer 4: I'm not sure how to implement basically everything here other than labels.  Ginkgo doesn't provide any facility to add metadata to tests other than labels.

We could write some helpers that apply metadata in a particular way, but it'd probably have to all be stuffed into labels - unless you have another idea?

Reviewer 4: TBD but this can be resolved.  We can handle it by walking the specs in injecting the data, or labeling the tests.

test cases developing a dependence on one another, `run-test` must make no
assumption about the number or composition of tests it is asked to run
(other than they will be consistent with the test environment and isolation
requirements defined by the "list" verb).

`openshift-tests` will randomize tests into multiple different parallel
executions of the extension binary in order to amortize initialization
costs.

Standard output from the invocation will be serialized into JSONL with
each line associated with a single test run result.

```python
{ "name": "T1", "result": "success", }
{ "name": "T1", "result": "success", }
{ "name": "T1", "result": "success", }
```

A single test result will contain the following information, encoded on a single line:
```python
{
    "name": "test name",
    
    # Duration of test run in milliseconds
    "duration": "1012",
    
    # The outcome of the test; pass, fail, skip, timeout.
    "result": "pass",
    "output": "standard output of the test",
    "error": "standard error of the test",
    
    # Human-readable messages to further explain 
    # skips / timeouts / etc.
    "messages": [
        "example message"
    ]
}

> ```

Reviewer 4: Suggest some additional fields:

- startTime
- endTime

Author: Good call. Adding in the next push.

Author: Origin will decorate the results with the test spec in the aggregated output. So we will ultimately have this information in the DB.

Reviewer 4: Should `ExtensionTestResult` just include the whole `ExtensionTestSpec`, so the results from invoking it from an extension binary produces the same output as if it was run from origin and aggregated?

Author: It will never be precisely the same. For example, only `origin` knows the full environment (e.g. context & full versions). There are pros/cons to including the spec in the result:
1. Con: The amount of output a human user of an extension binary must read through increases.
2. Pro: origin does not need to do the work of adding spec+result. 
3. Con: #2 is only true so long as the extension output is consistent with exactly what origin outputs in the aggregated format.  For example, if origin wants to move `component` to another place in the aggregated schema, it is free to do so today without having to remove anything from the extension result output from older versions of the extension. 
4. Pro: Consistency between origin and extension outputs.

The cons seemed to outweigh the pros, so I'm ever so slightly on the "bare minimum" result side.

The `update` verb will create or update files under `hack/.openshift-tests-exension/product/type/component/*`, by default
(basedir defaults to `hack/.openshift-tests-extension`).
If an incompatible change is introduced from the prior invocation of `update` (e.g. changing 
a test name without preserving the original), `update` will raise an error 
which the component owner must correct before committing their change in git.

The content of the files stored under `.openshift-tests-extension` is subject to change and wholly
defined by the `github.com/openshift-eng/openshift-tests-extension` module. Individual component
owners should not modify files under this path except through invocations of `update`.

#### Test Result Aggregation

`openshift-tests` will store test results in the CI artifacts repository in a JSONL file.
The output will be similar to that of the `run-test` output, but each test result 
will be expanded to include:
- Component & subcomponent
- Environment
- Testing Context

The goal here is to allow a single file to contain comprehensive information necessary for
a system like Component Readiness to analyze it without, for example, needing to draw information

> from prowjob names.

Reviewer 4: We'll either need to produce junit xml for visualization in spyglass, or have spyglass support the new format.  I almost wonder if we shouldn't break this off into its own enhancement and come up with a modern spec for test results (https://xkcd.com/927/).  Or see if an existing standard fits our needs.

Author: Since the output of `run-tests` is a super-set of the existing junit XML (and origin aggregates and writes both files), we should be able to populate the old and new formats for as long as necessary for a smooth transition. TRT can migrate to the richer metadata source when they have time. Migrating spyglass is optional until we have a use case for the additional metadata.

                "openshift/conformance"
            ],

            # Test cases can be advertised from across a number
            # of extension binaries. The following cel expressions
            # are OR'd together. If they select a test advertised
            # from another binary, they will be considered part of
            # this advertised suite.
            # Test tags, names, and other attributes will be
            # described later.
            "qualifiers": [
                "(test.tags.suite==\"fips\" || test.name.contains(\"fips\")) && !test.labels.has(\"Disruptive\")",
                # Normally, CEL expressions evaluate against ALL test
                # specs discovered by origin, regardless of the extension
                # that advertised them. This makes it easy for 
                # a suite defined in one extension to "absorb" tests 
                # specified by another extension. To limit the expression
                # to only tests in this extension, "owner" will be true
                # only when evaluating tests advertised from this 
                # extension.
                # "owner" is equivalent to "extension.component == self.component"

>                 "owner && test.name.contains(\"FIPS\"))"

Reviewer 4: I'm not sure how to wire up the `owner` boolean value without a lot of complexity from origin.  Could we use `source` instead (equal to the component identifier)?

Example:

```golang
	ext.AddSuite(
		e.Suite{
			Name:    "example/tests",
			Parents: []string{"openshift/conformance/parallel"},
			Qualifiers: []string{
				`source == "openshift:payload:example-tests" `,
			},
		})
```

It also has the benefit of allowing opting in to some component's tests, but not all.

Author: The intent here was simplicity for the suite-writer. You solved this with the AddSuite vs AddGlobalSuite, so we drop 'owner'.

costs.

Standard output from the invocation will be serialized into JSONL with
each line associated with a single test run result.

```python
{ "name": "T1", "result": "success", }
{ "name": "T1", "result": "success", }
{ "name": "T1", "result": "success", }
```

A single test result will contain the following information, encoded on a single line:
```python
{
    "name": "test name",
    
    # RFC-3339 start and stop time in UTC with millisecond precision.
    "startTime": "2026-01-02T15:04:05.000Z",
    "endTime": "2026-01-02T15:04:06.840Z",
    
    # The outcome of the test; pass, fail, skip, timeout.
    "result": "pass",
    "output": "standard output of the test",
    "error": "standard error of the test",
    
    # Human-readable messages to further explain 
    # skips / timeouts / etc.
    "details": [
        "example message"
    ]
}
```

When developing an extension, it may be useful to be able to run a suite
advertised by an extension directly. This can be accomplished using a
combination of `list` and `run-test`.

```
$ extension list --suite my-suite --platform aws ... | extension run-test --platform aws ...  --list --

> ```

Reviewer 4: I think I prefer having a native run-suite, too.  OTE has dumb one.

For this the syntax I ended up on:

```bash
$ extension list -o names --suite my-suite --platform aws ... | ./example-tests run-test --platform aws ...
```

Note that `run-test` will blindly execute tests in the list as quickly as possible,

> in parallel, without consideration for system resources or parallelism constraints

Reviewer 4: Just a note, OTE doesn't do parallel execution yet.  There's some oddities about how we invoke ginkgo tests today.  I think I just haven't found all the things I need mutexes for yet.

Reviewer 4: You can resolve this one, OTE executes in parallel.  Ginkgo has a mutex to force serial execution but other frameworks would be parallelized.

(basedir defaults to `hack/.openshift-tests-extension`).
If an incompatible change is introduced from the prior invocation of `update` (e.g. changing 
a test name without preserving the original), `update` will raise an error 
which the component owner must correct before committing their change in git.

The content of the files stored under `.openshift-tests-extension` is subject to change and wholly
defined by the `github.com/openshift-eng/openshift-tests-extension` module. Individual component
owners should not modify files under this path except through invocations of `update`.

#### Extension Implementation
github.com/openshift-eng/openshift-tests-extension will make contributing tests as simple
as possible. An example of what implementing the extension interface with this module
is provided below.

```golang
componentExtension := extensions.NewExtension("openshift", "payload", "hyperkube")

// boilerplate to initialize ginkgo omitted

// Allow extensions to Walk tests and return them wrapped
// in a wrapper extensions.ExtensionTestSpec. We can offer these
// helper functions to ingest tests from different testing framework.

> var ginkgoExtensionSpecs extensions.ExtensionTestSpecSet  // an abstraction around a group of ExtensionTestSpec

Reviewer 4: I used ExtensionTestSpecs, which is consistent with ginkgo's plural collection of objects (e..g Label/Labels)

// In this case, they are asking us to wrap Ginkgo tests they assert follow our 
// historical conventions.

> ginkgoExtensionSpecs, err = extensions.BuildExtensionTestSpecsFromOpenShiftGinkgoSuite(ginkgo.GetSuite())

Reviewer 4: `ginkgo.GetSuite()` isn't needed, there's more initialization needed, which BETSFOGS deals with

// similar story with tags.
customExtensionSpecs.Tag({"mytag": "myvalue"})  

// Here, OpenShift convention was used to determine which tests to run on AWS
// As a contrived example, let's say we want to run these same tests on GovCloud
awsTests := ginkgoExtensionSpecs.Select(extensions.PlatformSelector(platforms.AWS))
awsTests.AddPlatform(platforms.GovCloud)

// And so on with expressive grouping & metadata management. awsTests
// references the same objects as ginkgoExtensionSpecs, so changes in
// one are reflected immediately in both sets.
awsTests.Select(extensions.LabelSelector(platform.SLOW)).SetResource(resource.CPU, "5000m")

// Since this set of tests doesn't use the OpenShift naming conventions, they
// must constrain environmental parameters themselves.
customExtensionSpecs.Select(extensions.NameRegExSelector(".*cld/aws.*")).SetPlatforms(platforms.AWS)

// Since everything is an ExtensionTestSpec, you can mix sets.
allSpecsSet := customExtensionSpecs
allSpecsSet.Extend(ginkgoExtensionSpecs)

// The author can iterate through all tests and, for example,
// set OriginalName for each based on a configuration file
// or dictionary maintained in their repository.

> allSpecsSet.Walk(func(*ExtensionTestSpec))

Reviewer 4: ```suggestion
allSpecsSet.Walk(func(*ExtensionTestSpec) {
   // do stuff
})
```

// Finally, some or all of the specs are added for the component. 
componentExtension.AddTestSpecSets(allSpecsSet)

customSuite := componentExtension.DefineSuite("my-custom-suite")
customSuite.AddQualifier("(test.tags.mytag==\"myvalue\"")  // CEL expression which selects appropriate tests regardless of the extension which defines the test

> customSuite.AddParent("openshift/conformance") // Also run this suite when openshift/conformance runs

Reviewer 4: For now users just create a Suite struct, that seems more aligned with go not using excessive setters/getters

// The first component in the NewExtensionRegistry argument list
// will be selected by --component default, but also
// selectable with --component openshift:payload:hyperkube . 
// Most authors will be writing tests for one component, so origin
// running --component default, unless instructed otherwise, will "just work".
er := extensions.NewExtensionRegistry(componentExtension, ...)

// When using Cobra, this allows the main verb names to change without the author needing to
// enumerate them.

> root.AddCommand(er.CobraCommands()...)

Reviewer 4: This was a great suggestion.  Implemented as:

```
	root.AddCommand(cmd.DefaultExtensionCommands(registry)...)
```

        "source_url": "github.com/openshift/kubernetes"
    },

    # Versions of software detected on the system under
    # test. origin will combine the versions in the
    # aggregated test format output, so an extension
    # need only identify the version of the software
    # it is testing (e.g. an OLM operator need not
    # report the version of OpenShift).
    "versions": [
      {
        "name": "openshift",
        "version": "4.18.0",
        # A sha256 pullspec for the openshift payload
        # or a github.com URL to a specific commit. Something that
        # should completely identify the content of the 
        # software under test.
        "source": "...",
        
        # If the component was upgraded from a known version.
        "from": {
          "name": "...",
          "source": "..."
        }
      } 

>     ],

Reviewer 4: I'm surprised this included in `info`, given this is run early by origin I didn't expect it to require a kubeconfig or do anything other than provide info about itself and the tests extension.  Same for cloud driver down below.

These seem more like artifacts origin should generate (and it does at the end, there's a cluster-data.json from every run).

Author: If we can expand this framework to optional operators / service offering environments, `origin` will not know the versions or how to ascertain them. That said, `origin` will probably incorporate the OCP payload version into versions **itself**, so this wasn't a great example of why it is important.

                "source = \"openshift:payload:hyperkube\" && test.name.contains(\"FIPS\"))"
            ]
        }
    ],

    # Monitors are processes that will be run by origin for the duration of
    # test execution. They are similar to tests in that they can write
    # artifacts and report back tests results. They differ in that they
    # cannot have any conflicts/isolations (they must be able to run
    # during all testing).
    "monitors": [
        {
            # The name of the monitor (will be passed to run-monitor) 
            "name": "fips-endpoints",
            "description": "optional description",

            # Monitors can specify whether they should be run in order to 
            # save resources when they have no value. If they select any tests
            # origin identifies execution, origin will run the monitor. If
            # no qualifiers are specified, the monitor will always be run.
            "qualifiers": [
                "source = \"openshift:payload:hyperkube\" && test.name.contains(\"FIPS\"))"
            ]

>         }

Author: Moving into "list monitors". But leaving in the CEL, pending further discussion, as a way to further reduce the monitors if they are not needed based on the tests that will be run.

### Risks and Mitigations

#### Binary Incompatibility

Copying a binary from a different image into a running pod exposes us to binary incompatibility.

##### CPU Architecture

An arm64 binary will not run in an amd64 based pod. `openshift-tests` will need to run on
a build farm node with a CPU architecture consistent with architecture of the cluster
payload under test.

#### Runtime Size / Speed

Statically linked binaries are large relative to dynamically linked binaries. As the number
of extension binaries increase, so will the required tmpfs space used by pods running the
tests as those binaries are stored locally within the pod.

To mitigate this risk, we need to be aware of it, overprovision test pods with respect to
the memory normal test execution requires, and also include a test failure that will
TRT attention if the pod is using > 80% of allocatable memory.

#### Image Size

Statically linked binaries are relatively large. As the number of test extension binaries
increases, it may lead to a noticeable increase in overall payload size.

#### Poor Extension Implementation

Providing a contract that must be implemented without by test extensions without centralized

> control & review of the implementation leaves space for contributions that violate the

Author: Next push.
